cpp/include/raft/sparse/matrix/detail/preprocessing.cuh

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/device_coo_matrix.hpp>
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/core/resource/thrust_policy.hpp>
+#include <raft/linalg/map_reduce.cuh>
+#include <raft/matrix/init.cuh>
+#include <raft/sparse/convert/coo.cuh>
+#include <raft/sparse/neighbors/cross_component_nn.cuh>
+#include <raft/sparse/op/sort.cuh>
+
+#include <thrust/reduce.h>
+
+namespace raft::sparse::matrix::detail {
+
+/**
+ * This function creates a representation of input data (rows) that identifies when
+ * value changes in the input data. This function assumes data is sorted.
+ *
+ * @param[in] rows
+ *   The input data
+ * @param[in] nnz
+ *   The size of the input data.
+ * @param[in] counts
+ *   The resulting representation of the index value changes of the input. Should be
+ *   the same size as the input (nnz)
+ */
+__global__ void _scan(int* rows, int nnz, int* counts)
```

Comment:
I recall this keyword will cause the error in the CI, may need to change to `DI`.